File: code/python/1382.py
# https://leetcode.com/problems/balance-a-binary-search-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        self.arr = []

        def inorderTraversal(root):
            if not root: return
            inorderTraversal(root.left)
            self.arr.append(root.val)
            inorderTraversal(root.right)

        def sortedArrayToBST(l, r):
            if r < l: return None
            mid = l + (r - l) // 2
            node = TreeNode((self.arr[mid]))
            node.left = sortedArrayToBST(l, mid - 1)
            node.right = sortedArrayToBST(mid + 1, r)
            return node

        inorderTraversal(root)
        return sortedArrayToBST(0, len(self.arr) - 1)

File: code/cpp/1382.cpp
// https://leetcode.com/problems/balance-a-binary-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> arr{};

    void inorderTraversal(TreeNode* root) {
        if (!root) return;
        inorderTraversal(root->left);
        arr.emplace_back(root->val);
        inorderTraversal(root->right);
    }

    TreeNode* sortedArrayToBST(int l, int r) {
        if (r < l) return NULL;
        int mid = l + (r - l) / 2;
        TreeNode* node = new TreeNode((arr[mid]));
        node->left = sortedArrayToBST(l, mid - 1);
        node->right = sortedArrayToBST(mid + 1, r);
        return node;
    }

    TreeNode* balanceBST(TreeNode* root) {
        inorderTraversal(root);
        return sortedArrayToBST(0, arr.size() - 1);
    }
};
